Issue/2361 prune non premium

e-carlin opened the pull request:
Fix #2367, #1888 

Part of #2360 
- revive pruning pkcli
- Differentiate between premium and non-premium users

e-carlin commented on sirepo/pkcli/admin.py:
```diff
@@ -88,33 +88,42 @@ def purge_guest_users(days=180, confirm=False):
+        r = {}
+        for d in pkio.sorted_glob(simulation_db.user_dir_name().join('*')):
+            u = simulation_db.uid_from_dir_name(d)
+            # TODO(e-carlin): When does a path end in /src?
```
@robnagler when does this happen?

robnagler replied:
@moellep I always forget where this code is. We have a "src" user.

robnagler commented on sirepo/job_supervisor.py:
```diff
@@ -509,7 +514,11 @@ def _ops_to_cancel():
+            r = await self._receive_api_runStatus(req)
+            if r.state == job.MISSING:
+                # happens when the run dir is deleted w/o us knowing (ex admin cli deletion)
+                return await self._receive_api_runSimulation(req)
```
This is infinitely recursive so we need some sort of counter.

robnagler replied:
I don't see this resolved.

robnagler commented on sirepo/job_supervisor.py:
```diff
@@ -509,7 +514,11 @@ def _ops_to_cancel():
-            return await self._receive_api_runStatus(req)
+            r = await self._receive_api_runStatus(req)
+            if r.state == job.MISSING:
+                # happens when the run dir is deleted w/o us knowing (ex admin cli deletion)
```
`ex pkcli.admin.purge_basic_users` so a reader can find it. `admin cli deletion` is a bit too cryptic.

robnagler replied:
I don't see this change.

robnagler commented on sirepo/job_supervisor.py:
```diff
@@ -691,7 +700,15 @@ def _req_is_valid(self, req):
                     o.send()
-                    return await o.reply_get()
+                    r =  await o.reply_get()
+                    # POSIT: All non runSimulation jobs that could run into
```
runSimulation or here are the only places we need to check for runDirNotFound.

e-carlin replied:
Can you phrase what your saying a different way? I'm not sure what you're saying.

robnagler replied:
The POSIT is that we know all the places that runDirNotFound needs to be caught. There are two: runSimulation and this location. "All non runSimulation" implies that runSimulation also needs to catch runDirNotFound. I thought it should be more explicit.

Probably being too pedantic.

e-carlin replied:
Thanks. I rephrased it. Thoughts?

robnagler commented on sirepo/job_supervisor.py:
```diff
@@ -691,7 +700,15 @@ def _req_is_valid(self, req):
+                    r =  await o.reply_get()
+                    # POSIT: All non runSimulation jobs that could run into
+                    # runDirNotFound call _send_with_single_reply()
+                    if r.get('runDirNotFound'):
```
reverse the test so that the longer clause is less indented.

robnagler commented on sirepo/pkcli/admin.py:
```diff
@@ -77,44 +77,53 @@ def move_user_sims(target_uid=''):
 
-def purge_guest_users(days=180, confirm=False):
-    """Remove old users from db which have not registered.
+def purge_basic_users(days=180, confirm=False):
```
basic => free?

robnagler commented on sirepo/pkcli/admin.py:
```diff
@@ -77,44 +77,53 @@ def move_user_sims(target_uid=''):
+            else:
+                r[d] = u
+        if confirm:
+            pkio.unchecked_remove(*r.keys())
```
we want to purge simulation output, but not the actual user.

robnagler commented on sirepo/pkcli/job_agent.py:
```diff
@@ -253,18 +253,25 @@ def terminate(self):
+            elif msg.jobCmd == 'fastcgi':
+                c = _FastCgiCmd
+            p = c(msg=msg, dispatcher=self, op_id=msg.opId, **kwargs)
+            if msg.jobCmd == 'fastcgi':
```
for clarity, I think the `except` should be above this line. There's some global state being modified, and those operations should be atomic (or in error, i.e. exception raised). I would also move the `c` math into a function. It will clean up the code a little.

robnagler commented on sirepo/pkcli/job_agent.py:
```diff
@@ -321,6 +328,8 @@ def _fastcgi_accept(self, connection, *args, **kwargs):
             )
             # last thing, because of await: start fastcgi process
             await self._cmd(m, send_reply=False)
+        if not pkio.py_path(msg.runDir).exists():
```
would this not want to be first, before starting fastcgi command?

robnagler commented on sirepo/pkcli/job_agent.py:
```diff
@@ -783,11 +793,20 @@ def start(self):
 
+    def _assert_run_dir_exists(self):
+        if self.cmd.msg.jobCmd not in ('prepare_simulation', 'compute') \
+           and not self.cmd.run_dir.exists():
```
This part could be shared with the other `raise` above. Not critical, but would make the code simpler:

```
if self.cmd.msg.jobCmd not in ('prepare_simulation', 'compute'):
    _assert_run_dir(self.cmd.run_dir)
```

robnagler commented on tests/pkcli/admin_test.py:
```diff
@@ -10,66 +10,61 @@
-        None,
-        '{}: expecting uid to deleted from db', res
+    u = fc.sr_login_as_guest()
+    sirepo.auth_db.UserRole.delete_roles(u, [sirepo.auth.ROLE_PREMIUM])
```
how about login as email and then there's no need to reach inside the abstraction that guest always has a ROLE_PREMIUM.

e-carlin commented:
@robnagler don't review yet. Still working on a few of your comments.

e-carlin commented:
@robnagler ready for another review. All comments addressed except [this one](https://github.com/radiasoft/sirepo/pull/2394#discussion_r398222015).

robnagler commented on sirepo/job_supervisor.py:
```diff
@@ -284,16 +284,21 @@ def __db_init(self, req, prev_db=None):
-            jobRunMode=c.jobRunMode,
+            jobRunMode=c.get(
+                'jobRunMode',
+                self.db.jobRunMode if 'db' in self else None,
```
This is a little funky and deserves a comment, perhaps.

e-carlin replied:
I added an assert and comment.

e-carlin replied:
And changed the behavior slightly. After more thought we should never run into the None case.

robnagler commented on sirepo/job_supervisor.py:
```diff
@@ -691,7 +700,15 @@ def _req_is_valid(self, req):
+                        self.__db_init(req, prev_db=self.db)
+                        assert self.db.status == job.MISSING, \
+                            'expecting missing status={}'.format(self.db.status)
+                        return PKDict(state=job.MISSING)
```
self.db.status, not job.MISSING. Keeps the encapsulation.

e-carlin commented:
> Am I supposed to review? I did but maybe not ready?

Yes. I think it was ready. It looked like you possibly were commenting on outdated code? 
I believe 69ae657  and   93b96a3 addressed your comments up to that point. In any event all comments have been addressed now and ready for another pass.

robnagler commented on sirepo/job_supervisor.py:
```diff
@@ -284,16 +284,29 @@ def __db_init(self, req, prev_db=None):
+        r = c.get('jobRunMode')
+        if not r:
+            assert c.api != 'api_runSimulation', \
+                'api_runSimulation must have a jobRunMode'
```
provide context eg. printing `c` in the error message.

robnagler commented on sirepo/job_supervisor.py:
```diff
@@ -509,7 +522,13 @@ def _ops_to_cancel():
+            r = await self._receive_api_runStatus(req)
+            if r.state == job.MISSING:
+                # happens when the run dir is deleted w/o us knowing (ex pkcli.admin.purge_free_users)
+                assert not called_from_self, \
```
`called_from_self` is a bit confusing. `self` is a meaningful concept, and it's a parameter on the call. Rather have it be a counter: recursion_depth, recursion_level, call_depth, etc. When the counter exceeds a threshold, it should throw an error. In this case, `1` is an appropriate threshold.

robnagler commented on sirepo/job_supervisor.py:
```diff
@@ -691,7 +710,16 @@ def _req_is_valid(self, req):
+                    if not r.get('runDirNotFound'):
+                        return r
+                    self.__db_init(req, prev_db=self.db)
+                    assert self.db.status == job.MISSING, \
```
This particular assertion is ok for now. I'm not sure we want it in the long run. At some point we have to trust the code to do what it is supposed to do. We aren't programming Eiffel. We have tests for this type of validation.

robnagler commented on sirepo/pkcli/admin.py:
```diff
@@ -77,44 +77,60 @@ def move_user_sims(target_uid=''):
 
-def purge_guest_users(days=180, confirm=False):
-    """Remove old users from db which have not registered.
+def purge_free_users(days=180, confirm=False):
```
Not exactly what's happening: expunge_free_simulations.

robnagler commented on sirepo/pkcli/admin.py:
```diff
@@ -77,44 +77,60 @@ def move_user_sims(target_uid=''):
-def purge_guest_users(days=180, confirm=False):
-    """Remove old users from db which have not registered.
+def purge_free_users(days=180, confirm=False):
+    """Remove basic users simulations
```
Remove expired simulation results for non-premium users

robnagler commented on sirepo/pkcli/admin.py:
```diff
@@ -77,44 +77,60 @@ def move_user_sims(target_uid=''):
+        u = _get_premium_uids()
+        n = srtime.utc_now()
+        res = []
+        for p in pkio.sorted_glob(simulation_db.user_dir_name().join(
```
Let's add this abstraction to simulation_db, which controls the structure of simulation run dirs.

robnagler commented on sirepo/pkcli/admin.py:
```diff
@@ -77,44 +77,60 @@ def move_user_sims(target_uid=''):
 
+    def _check_expired_run_dir(path):
+        d = str(path).lower()
+        if 'report' in d or 'animation' in d:
```
I don't see why this code needs to concern itself with the simulation run directory name.

robnagler commented on sirepo/pkcli/admin.py:
```diff
@@ -77,44 +77,60 @@ def move_user_sims(target_uid=''):
+    def _check_expired_run_dir(path):
+        d = str(path).lower()
+        if 'report' in d or 'animation' in d:
+            for f in pkio.walk_tree(path):
```
I don't think this is necessary, since run directories are re-written every time we can just rely on the ctime for the run_dir.

robnagler commented on sirepo/pkcli/admin.py:
```diff
@@ -77,44 +77,60 @@ def move_user_sims(target_uid=''):
+            else:
+                res.append(str(path))
+
+    def _get_premium_uids():
```
Encapsulate this in auth_db.

I think, perhaps, auth_db should defined ROLES as PKDict(PREMIUM='premium', ...).

robnagler commented on sirepo/pkcli/admin.py:
```diff
@@ -77,44 +77,60 @@ def move_user_sims(target_uid=''):
-        auth_db.UserRegistration.delete_all_for_column_by_values('uid', dirs_and_uids.values())
-
-    return dirs_and_uids
+    with auth_db.thread_lock:
```
This is insufficient and problematic. The problem is that Flask creates run directories without checking auth_db thread_lock, and in any event, flask is running in a different process than this pkcli so thread locks doesn't work.

We need to think about if locking is necessary when removing run directories. 

Since the supervisor controls run directories. We could send it a list of run directories to purge from this code.

The advantage of this is that the supervisor could actually store a different job state when a simulation is purged instead of missing. It would mean not hitting runDirNotFound on this case, that is, the supervisor would see the new state (e.g. job.PURGED_BASIC) and it would be able to return this to the GUI, which could then explain to the user: your simulations were purged, and you might want to upgrade to premium so this doesn't happen.

robnagler commented on sirepo/pkcli/job_agent.py:
```diff
@@ -208,6 +213,14 @@ def terminate(self):
+    def _get_cmd_type(self, msg):
+        c = _Cmd
+        if msg.jobRunMode == job.SBATCH:
+            c = _SbatchRun if msg.isParallel else _SbatchCmd
```
`return`, don't assign, then the code is cleaner:

```py
if msg.jobRunMode == job.SBATCH:
     return _SbatchRun if msg.isParallel else _SbatchCmd
elif msg.jobCmd == 'fastcgi':
     return _FastCgiCmd
return _Cmd
```

Easier to follow the logic, especially for the default case

e-carlin commented:
After discussion with @robnagler we decided that we're going to have a periodic callback in the supervisor that does the pruning at 2am every day. Closing while I work on that.
